### PR TITLE
PEM: Add more library context aware PEM readers

### PR DESCRIPTION
**Pull request by levitte**
PEM_read_bio_PUBKEY_ex() and PEM_read_bio_Parameters_ex() are added to
complete PEM_read_bio_PrivateKey_ex().  They are all refactored to be
wrappers around the same internal function.

This also includes a modification of OSSL_STORE so it can report public keys separately from private keys.

### CODE REVIEW

**Comment by levitte** [comment]
#12587 will depend on this

**Comment by levitte** [comment]
> #12587 will depend on this

#12512 as well, come to think of it...

File: crypto/pem/pem_pkey.c [hunk 1]
```diff
@@ -65,11 +90,57 @@ EVP_PKEY *PEM_read_bio_PrivateKey_ex(BIO *bp, EVP_PKEY **x, pem_password_cb *cb,
 
  err:
     OSSL_STORE_close(ctx);
-    UI_destroy_method(ui_method);
+    UI_destroy_method(allocated_ui_method);
     OSSL_STORE_INFO_free(info);
     return ret;
 }
 
+EVP_PKEY *PEM_read_bio_PUBKEY_ex(BIO *bp, EVP_PKEY **x,
```
Comment (slontis) [inline crypto/pem/pem_pkey.c:98]: Another case where it will be changed to with_libctx in another PR.

File: crypto/pem/pem_pkey.c [hunk 2]
```diff
@@ -65,11 +90,57 @@ EVP_PKEY *PEM_read_bio_PrivateKey_ex(BIO *bp, EVP_PKEY **x, pem_password_cb *cb,
...
+{
+    BIO *b;
+    EVP_PKEY *ret;
+
+    if ((b = BIO_new(BIO_s_file())) == NULL) {
+        PEMerr(0, ERR_R_BUF_LIB);
+        return 0;
+    }
+    BIO_set_fp(b, fp, BIO_NOCLOSE);
+    ret = PEM_read_bio_PrivateKey_ex(b, x, cb, u, libctx, propq);
```
Comment (slontis) [inline crypto/pem/pem_pkey.c:125]: Did you mean to use Private here??

File: crypto/store/store_lib.c [hunk 3]
```diff
@@ -422,7 +448,7 @@ EVP_PKEY *OSSL_STORE_INFO_get1_PKEY(const OSSL_STORE_INFO *info)
         return info->_.pkey;
     }
     OSSL_STOREerr(OSSL_STORE_F_OSSL_STORE_INFO_GET1_PKEY,
-                  OSSL_STORE_R_NOT_A_KEY);
+                  OSSL_STORE_R_NOT_A_PRIVATE_KEY);
```
Comment (slontis) [inline crypto/store/store_lib.c:451]: Didnt the code above say this could be a public key also? So I think the old error is better?

File: doc/man3/OSSL_STORE_INFO.pod [hunk 4]
```diff
@@ -3,15 +3,17 @@
...
-OSSL_STORE_INFO_get0_NAME_description, OSSL_STORE_INFO_get0_PARAMS,
+OSSL_STORE_INFO_get0_NAME_description,
+OSSL_STORE_INFO_get0_PARAMS, OSSL_STORE_INFO_get0_PUBKEY,
 OSSL_STORE_INFO_get0_PKEY, OSSL_STORE_INFO_get0_CERT, OSSL_STORE_INFO_get0_CRL,
 OSSL_STORE_INFO_get1_NAME, OSSL_STORE_INFO_get1_NAME_description,
-OSSL_STORE_INFO_get1_PARAMS, OSSL_STORE_INFO_get1_PKEY,
-OSSL_STORE_INFO_get1_CERT,
-OSSL_STORE_INFO_get1_CRL, OSSL_STORE_INFO_type_string, OSSL_STORE_INFO_free,
+OSSL_STORE_INFO_get1_PARAMS, OSSL_STORE_INFO_get1_PUBKEY,
+OSSL_STORE_INFO_get1_PKEY, OSSL_STORE_INFO_get1_CERT, OSSL_STORE_INFO_get1_CRL,
```
Comment (slontis) [inline doc/man3/OSSL_STORE_INFO.pod:11]: Each space at end of line?

File: doc/man3/OSSL_STORE_INFO.pod [hunk 5]
```diff
@@ -76,23 +81,25 @@ STORE_INFO_get_type_string() takes a STORE type number and returns a
...
 OSSL_STORE_INFO_get0_NAME(), OSSL_STORE_INFO_get0_NAME_description(),
-OSSL_STORE_INFO_get0_PARAMS(), OSSL_STORE_INFO_get0_PKEY(),
-OSSL_STORE_INFO_get0_CERT() and OSSL_STORE_INFO_get0_CRL() all take a
-B<OSSL_STORE_INFO> and return the held object of the appropriate OpenSSL
-type provided that's what's held.
+OSSL_STORE_INFO_get0_PARAMS(), OSSL_STORE_INFO_get0_PUBKEY(),
+OSSL_STORE_INFO_get0_PKEY(), OSSL_STORE_INFO_get0_CERT(),
+OSSL_STORE_INFO_get0_CRL()
+all take a B<OSSL_STORE_INFO> and return the held object of the appropriate
+OpenSSL type provided that's what's held.
```
Comment (slontis) [inline doc/man3/OSSL_STORE_INFO.pod:88]: 'provided that's what's held' should maybe be replaced with something like
'or NULL if the object does not exist'..

File: doc/man3/OSSL_STORE_INFO.pod [hunk 6]
```diff
@@ -76,23 +81,25 @@ STORE_INFO_get_type_string() takes a STORE type number and returns a
...
 
 OSSL_STORE_INFO_get1_NAME(), OSSL_STORE_INFO_get1_NAME_description(),
-OSSL_STORE_INFO_get1_PARAMS(), OSSL_STORE_INFO_get1_PKEY(),
-OSSL_STORE_INFO_get1_CERT() and OSSL_STORE_INFO_get1_CRL() all take a
-B<OSSL_STORE_INFO> and return a duplicate of the held object of the
+OSSL_STORE_INFO_get1_PARAMS(), OSSL_STORE_INFO_get1_PUBKEY(),
+OSSL_STORE_INFO_get1_PKEY(), OSSL_STORE_INFO_get1_CERT() and
+OSSL_STORE_INFO_get1_CRL()
+all take a B<OSSL_STORE_INFO> and return a duplicate of the held object of the
 appropriate OpenSSL type provided that's what's held.
```
Comment (slontis) [inline doc/man3/OSSL_STORE_INFO.pod:95]: same here..

File: doc/man3/PEM_read_bio_PrivateKey.pod [hunk 7]
```diff
@@ -67,8 +67,14 @@ PEM_write_bio_PKCS7, PEM_write_PKCS7 - PEM routines
                                    char *kstr, int klen,
                                    pem_password_cb *cb, void *u);
 
+ EVP_PKEY *PEM_read_bio_PUBKEY_ex(BIO *bp, EVP_PKEY **x,
```
Comment (slontis) [inline doc/man3/PEM_read_bio_PrivateKey.pod:70]: Need to run doc-nits on this..
They are not in the NAME section

**Review by slontis — COMMENTED** [review]
Travis failures due to doc-nits

File: crypto/pem/pem_pkey.c (hunk 2 above)
Comment (levitte) [reply crypto/pem/pem_pkey.c:125]: Ah, no

File: doc/man3/OSSL_STORE_INFO.pod (hunk 4 above)
Comment (levitte) [reply doc/man3/OSSL_STORE_INFO.pod:11]: Huh?  Oh, do you mean one name per line?

Comment (slontis) [reply doc/man3/OSSL_STORE_INFO.pod:11]: reworded in english..

File: crypto/store/store_lib.c (hunk 3 above)
Comment (levitte) [reply crypto/store/store_lib.c:451]: The loader knows how to distinguish a public key (has public parts but no private parts) and a private key (also has private parts), so while *the rest of OpenSSL* has no way of distinguishing those, the application may still need to know, so OSSL_STORE now tells them apart, just like it has told apart an EVP_PKEY with just parameters.  See the message in `OSSL_STORE_INFO_get1_PARAMS`...

Comment (levitte) [reply crypto/store/store_lib.c:451]: Ah, you're referring to the comment in `load_key_cert_crl()` (`apps/lib/apps.c`), aren't you?  Maybe I need to elaborate on that one.

File: doc/man3/OSSL_STORE_INFO.pod (hunk 5 above)
Comment (levitte) [reply doc/man3/OSSL_STORE_INFO.pod:88]: In that case, 'or NULL if the B<OSSL_STORE_INFO> holds an object of a different type'

File: doc/man3/PEM_read_bio_PrivateKey.pod (hunk 7 above)
Comment (levitte) [reply doc/man3/PEM_read_bio_PrivateKey.pod:70]: Fixed

File: crypto/pem/pem_pkey.c (hunk 2 above)
Comment (levitte) [reply crypto/pem/pem_pkey.c:125]: Fixed

File: doc/man3/OSSL_STORE_INFO.pod (hunk 5 above)
Comment (levitte) [reply doc/man3/OSSL_STORE_INFO.pod]: I rephrased it a bit more

File: doc/man3/OSSL_STORE_INFO.pod (hunk 4 above)
Comment (levitte) [reply doc/man3/OSSL_STORE_INFO.pod:11]: Ah.  No, there isn't.

File: crypto/store/store_lib.c (hunk 3 above)
Comment (levitte) [reply crypto/store/store_lib.c:451]: I've rewritten the comment.

File: doc/man3/OSSL_STORE_INFO.pod [hunk 8]
```diff
@@ -84,15 +84,17 @@ OSSL_STORE_INFO_get0_NAME(), OSSL_STORE_INFO_get0_NAME_description(),
...
+B<OSSL_STORE_INFO> type (as returned by OSSL_STORE_INFO_get_type())
+matches the function, otherwise NULL.
 
 OSSL_STORE_INFO_get1_NAME(), OSSL_STORE_INFO_get1_NAME_description(),
 OSSL_STORE_INFO_get1_PARAMS(), OSSL_STORE_INFO_get1_PUBKEY(),
 OSSL_STORE_INFO_get1_PKEY(), OSSL_STORE_INFO_get1_CERT() and
 OSSL_STORE_INFO_get1_CRL()
-all take a B<OSSL_STORE_INFO> and return a duplicate of the held object of the
-appropriate OpenSSL type provided that's what's held.
+all take a B<OSSL_STORE_INFO> and return a duplicate the object if
```
Comment (slontis) [inline doc/man3/OSSL_STORE_INFO.pod:95]: if holds if the

makes no sense :)

Comment (levitte) [reply doc/man3/OSSL_STORE_INFO.pod:95]: Ok, so please help me out.

For example, `OSSL_STORE_INFO_get1_PUBKEY()` will return non-NULL if `OSSL_STORE_INFO_get_type()` returns `OSSL_STORE_INFO_PUBKEY`, otherwise NULL (even if the **OSSL_STORE_INFO** holds on object, just of a different type).

Comment (levitte) [reply doc/man3/OSSL_STORE_INFO.pod:95]: Ohhhhhh, hold on, I didn't quite read what you wrote.  Should be "it holds it", of course...

Comment (levitte) [reply doc/man3/OSSL_STORE_INFO.pod:95]: Fixed

**Comment by levitte** [comment]
Ping (#12512 depends on this, so I'd rather not this be kept hanging)

**Review by t8m — APPROVED** [review]
LGTM

**Review by slontis — APPROVED** [review]
LGTM also

**Comment by levitte** [comment]
Merged

2274d22d39fd65b83c9c969cba86c8f03b3e0bec STORE: Distinguish public keys from private keys
6e5ccd58c85f05f8d47a1682125a207dbc694768 PEM: Add more library context aware PEM readers
5f2b7db09b6f9650ecf2b465ee9f737385623bd5 TEST: Use PEM_read_bio_PUBKEY_ex() and PEM_read_bio_PrivateKey_ex()
